lib/tokenize.cpp

```diff
@@ -10968,7 +10968,7 @@ void Tokenizer::simplifyAttribute()
             Token *functok = nullptr;
             if (Token::Match(after, "%name%|*")) {
                 Token *ftok = after;
-                while (Token::Match(ftok, "%name%|* !!("))
+                while (Token::Match(ftok, "%name%|::|<|>|* !!("))
```

Comment:
If there comes a `<` I have the feeling we should stop. If the code says ` foo<int>()`  I spontanously think the `functok` should be `foo`. It seems it would be preferable to use `ftok->link()` for the `<` but I am not sure if that is set yet. I fear you don't handle for instance `foo<int,int>()` well.

Reply:
That's not exactly the situation that caused me to add the `<` to this.  The problem that I ran into is that we have code like this in a header in our project:

```
template<T>
__attribute__((warn_unused_result))
Result<T>
member_method();
```

(you can see the actually problematic code at https://github.com/ros2/rclcpp/blob/9342c257b8a427663f62c05de853aa2ef1b78e6f/rclcpp/include/rclcpp/wait_set_template.hpp#L651-L654)

While removing the attribute, `simplifyAttribute` is looking to find the function name token by skipping over the return type.  But since `<` is (currently) considered an invalid token here, it just gives up at that point and throws a `syntaxError`.  That's why I added it to the list of "allowed" tokens to skip while looking for the actual name of the method.

That said, you are right; this code will not deal with multiple template arguments in the return value correctly.  I guess we have none of those in our codebase :).  I guess I could add `,` to the allowed list as well, but I'm not sure if we should keep adding tokens that we are allowed to skip or not.  Off-hand, I don't know of another approach to this, but if you have ideas I'm willing to try them out.

Reply:
> That said, you are right; this code will not deal with multiple template arguments in the return value correctly. I guess we have none of those in our codebase :). I guess I could add , to the allowed list as well, but I'm not sure if we should keep adding tokens that we are allowed to skip or not. Off-hand, I don't know of another approach to this, but if you have ideas I'm willing to try them out.

I would prefer that if a `<` is seen then the `Token::link()` is used. That normally points at the corresponding `>` token. If the `Token::link()` is nullptr then there is some method that will lookup the `>` but I don't remember from the top of my head which token that is. So the code should be something like:

    while (Token::Match(ftok, "%name%|::|<|* !!(")) {
        if (ftok->str() == "<")
            ftok = ftok->link();
        ftok = ftok->next();
    }

Reply:
> I would prefer that if a < is seen then the Token::link() is used. That normally points at the corresponding > token. If the Token::link() is nullptr then there is some method that will lookup the > but I don't remember from the top of my head which method that is.

Thanks for the pointer!  That was super helpful.  I think the API that you couldn't remember was `findClosingBracket`, and I've now implemented the loop to look for the `<`, try `->link`, and if that fails, fall back to `->findClosingBracket`.  I've also added a test to test multiple arguments inside the template brackets.

Please take a look at 53d0d4d when you have a chance.